frontend/src/hooks/useFormData/index.tsx

```diff
@@ -151,19 +165,52 @@ const useFormData = (): [FormData, Meta, Helpers] => {
     return [
         formData,
         { errorOnSubmit },
-        { handleInputChange, validateField, validateForm, handleSubmit },
+        { handleInputChange, validateField, handleSubmit },
     ];
 };
 
 export default useFormData;
 
-function toPayload(formData: FormData) {
-    return Object.fromEntries(
-        Object.entries(formData).map(([field, fieldData]) => [
-            field,
-            fieldData.value,
-        ])
-    );
+function isFieldMissing(field: keyof FormData, formData: FormData) {
+    const { value } = formData[field];
+    const isRequired = optionalFields.every((opF) => opF !== field);
+    const isEmpty =
+        value === null ||
+        value === undefined ||
+        value === "" ||
+        (Array.isArray(value) && !value.length);
+    return isEmpty && isRequired;
+}
+
+function noErrorsDetected(
+    errorActions: (() => void)[],
+    formData: FormData
+): formData is ValidFormData {
+    return !errorActions.length;
+}
+
+function toPayload(formData: ValidFormData): GameReportPayload {
+    return {
+        winner: formData.winner.value,
+        loser: formData.loser.value,
+        side: formData.side.value,
+        victory: formData.victory.value,
+        match: formData.match.value,
+        competition: formData.competition.value,
+        league: formData.league.value,
+        expansions: formData.expansions.value,
+        treebeard: formData.treebeard.value,
+        actionTokens: formData.actionTokens.value,
+        dwarvenRings: formData.dwarvenRings.value,
+        turns: formData.turns.value,
+        corruption: formData.corruption.value,
+        mordor: formData.mordor.value,
+        initialEyes: formData.initialEyes.value,
+        aragornTurn: formData.aragornTurn.value,
+        strongholds: formData.strongholds.value,
+        interestRating: formData.interestRating.value,
+        comment: formData.comment.value,
```

Comment:
Second issue: this block is gross and I want the Haskell `..` 😭  I tried to create a helper function like below to at least do the transformation to values first, which was only an initial step, and couldn't get even get that working:
```
type ReducedFormData = { [K in keyof FormData]: FormData[K]["value"] };

function transformValues(
    formData: FormData,
    transform: <K extends keyof FormData>(
        key: K,
        obj: FormData
    ) => FormData[K]["value"]
): ReducedFormData {
    return objectKeys(formData).reduce((accum, key) => {
        if (key in formData) {
            accum[key] = transform(key, formData);
        }
        return accum;
    }, {} as ReducedFormData);
}
```
Any ideas? And/or is it worth it to keep trying? The TS error is on `accum[key] = transform(key, formData);`, and it's:
```
Type 'string | number | boolean | ("League" | "Tournament")[] | ("LoME" | "WoME" | "KoME" | "Cities" | "Fate of Erebor" | "Treebeard")[] | ("Rivendell" | "Grey Havens" | "Helm's Deep" | ... 10 more ... | "Iron Hills (Fate of Erebor expansion only)")[] | null' is not assignable to type 'never'.
  Type 'null' is not assignable to type 'never'
```

Reply:
This is a sensible way of laying out the options, I think I will stick with the trade of more lines of code in exchange for keeping type safety! But agnostic of that you're right that `GameReportPayload` would be better as a mapped type, nice. Added that: [2620b0b](https://github.com/sirrus233/wotr-community-website/pull/28/commits/2620b0b25705e1e43ba64536e8ce07005ac31d8e)